Updated CRV branch names and added a PEs per layer variable.

sam-grant opened the pull request:
1.  Updated CRV branch names (following discussion with Andy, Ralf, and Yuri): 
- `crvhit` —> `crvcoincs`;
- `crvpulseinfo` —> `crvpulses`;
- `crvwaveforminfo` —> `crvdigis`.

2. Added a new leaf to `crvcoincs` tree: `float PEsPerLayer[4]`. This is an array of photoelectron counts indexed by layer number (four layers per module), which is filled by looping through the reco pulses for each coincidence cluster. See [Mu2e-doc-47954](https://mu2e-docdb.fnal.gov/cgi-bin/sso/ShowDocument?docid=47954). I don't see any change in the `TrkAna` run time with this additional `for` loop. 

3. As a side note, the original `PEs` leaf was changed from an `int` to a `float`, after discussing with Ralf and Yuri. The convention that `PEs` are integers is outdated.

AndrewEdmonds11 commented on fcl/prolog.fcl:
```diff
@@ -173,7 +173,7 @@ TrkAnaTreeMaker : {
   ProcessEmptyEvents : false
-  FillCRVHits : true
+  FillCRVCoincs : true
   FillCRVPulses : false
```
Should we also change ```FillCRVPulses``` to ```FillCRVDigis``` too?

sam-grant replied:
In the current scheme I think this comes down to preference. `FillCRVPulses` switches on both the `crvdigis` and `crvpulses` branches, since these are both filled in `CrvInfoHelper::FillCrvPulseInfoCollections`:

```
      if(_fillcrvpulses){
        _crvHelper.FillCrvPulseInfoCollections(_crvRecoPulses, _crvDigiMCs, _crvDigis,
                                               _crvpulses, _crvpulsesmc, _crvdigis);
      }
```

I'm not sure why we do it this way to be quite honest, you could split `digis` and `pulses` up and have two different flags. That might make things less confusing?

AndrewEdmonds11 replied:
Sounds good. If it's turns out to be too much hassle to separate it out, then I think I would prefer ```digis``` over ```pulses``` in the fcl parameter name. We're also thinking about updating the parameter names for the flags that the tracker uses for its digis so some consistency there would be good

sam-grant replied:
Should I just push what I've done on this comment even if I'm not sure if it's an improvement? Is it easy to remove commits from PRs? Thanks

kutschke replied:
To "undo" a commit you create the inverse commit and commit it.  There is a command to create the inverse commit.  So it's always there in the history but so long as you end up in the right place in the end, that's what you want.

AndrewEdmonds11 commented on inc/CrvHitInfoReco.hh:
```diff
@@ -32,7 +33,10 @@ namespace mu2e
-    {}
+    {
+      // Copy PEsPerLayer array elements
+      for (int i = 0; i < 4; i++) this->PEsPerLayer[i] = PEsPerLayer[i];
```
Is the number of CRV layers defined somewhere in the Offline? I'd prefer to use a named variable than a bare 4. Also, I prefer to have curly brackets for the for loop even if it's just one line for safety

sam-grant replied:
Good point. I had a quick look at there is a parameter `nLayers_` in `Offline/GeometryService/inc/CosmicRayShieldMaker.hh` which looks like a likely candidate. I will double check and add it in. I'll also add the curly brackets.

ehrlich-uva replied:
You can use `CRVId::nLayers` defined in `Offline/DataProducts/inc/CRVId.hh`.

sam-grant replied:
Thanks Ralf

kutschke replied:
Is it too late to define all of the c-style arrays of dimension 4  as std::array<int,4>?   Then you can do the copy in the intiailizer list:  PEsPerLayer(PEsPerLayer) .  One of the obscure but handy corners of c++ is that in an initializer list, the compiler knows that it means to have a data member and an argument with the same name

It it is too late definitely use std::array, then use CRVId::nLayers from Offline/DataProducts/inc/CRVId.hh  .  That's a compile time condstant  rather than a variable that you have to look up at run time.

I don't think that the recommendation of Offline/GeometryService/inc/CosmicRayShieldMaker.hh will work.  Do not try to make an instance of that class yourself.  It is to only be using insde the geometry service.

sam-grant replied:
Hi Rob. Thanks for the comment. It's not too late to define the array as a `std::array<int, CRVId::nLayers>` and then copy it to the initializer as `PEsPerLayer(PEsPerLayer)`, I will add that in. Ralf's recommendation to use CRVId::nLayers from `Offline/DataProducts/inc/CRVId.hh` works well, so I won't attempt to use anything from `CosmicRayShieldMaker.hh`.

sam-grant replied:
Although I suppose it should be an array of float rather than ints.

sam-grant replied:
It seems like you need to copy the array elements using curly brackets: `PEsPerLayer{PEsPerLayer}` rather than `PEsPerLayer(PEsPerLayer)`.

AndrewEdmonds11 commented on src/CrvInfoHelper.cc:
```diff
@@ -37,12 +37,34 @@ namespace mu2e
       const CrvCoincidenceCluster &cluster = crvCoincidences->at(i);
 
+      // Get the PEs per layer from the reco pulses
+      float PEsPerLayer_[4] = {0.}; // Four layers, each one initiliased to zero PEs in a static array
```
Similar comment to earlier about the bare 4

sam-grant replied:
I'll update this with the `nLayers` from the offline geometry.

ehrlich-uva replied:
Same as above: You can use `CRVId::nLayers` defined in `Offline/DataProducts/inc/CRVId.hh`.

AndrewEdmonds11 commented on src/CrvInfoHelper.cc:
```diff
@@ -37,12 +37,34 @@ namespace mu2e
+      float PEsPerLayerSum = 0.; 
+      for (size_t j=0; j<4; j++) PEsPerLayerSum += PEsPerLayer_[j];
+      float deltaPEs = abs(PEsPerLayerSum - cluster.GetPEs());
+      if (deltaPEs > 1e-3) std::cout<<"CrvInfoHelper warning: the sum of PEs per layer differs by " << deltaPEs << " from total PEs in the coincidence!"<<std::endl;
```
Is this warning bad enough that we should really throw an exception and stop execution?

sam-grant replied:
Maybe, although the `1e-3` limit is just something I made up, which is the main reason I made it a `cout` rather than a `cerr`. To be honest I don't understand why there is any difference between the two values at all. 

There is a tricky detail where you need to change the way you get your PEs from the reco pulses depending on the way they were reconstructed in the coincidence. There are a few different ways PEs are formed, see [Offline/CrvCoincidenceFinder_module.cc#L314-L317](https://github.com/Mu2e/Offline/blob/main/CRVReco/src/CrvCoincidenceFinder_module.cc#L314-L317), and I was originally using `GetPEs()` rather than `GetPEsNoFit()`, but was seeing a large (like 5-10 PEs) discrepancy between the total PEs from the coincidence and the sum of the PEs from the reco pulses, which is why I originally made that printout. 

Maybe we need to discuss this with Ralf?

AndrewEdmonds11 replied:
Yes, I think it would be good to discuss with Ralf. It would be good if TrkAna didn't need to know how the PEs were formed and could just use a ```GetPEs()``` function and be given the correct ones. Then we don't need to update anything if the method changes in the CRV reconstruction code

AndrewEdmonds11 commented on src/TrkAnaTreeMaker_module.cc:
```diff
@@ -157,8 +157,8 @@ namespace mu2e {
-        fhicl::Atom<bool> crvhits{Name("FillCRVHits"),Comment("Flag for turning on crv CoincidenceClusterbranches"), false};
-        fhicl::Atom<bool> crvpulses{Name("FillCRVPulses"),Comment("Flag for turning on crvpulseinfo(mc), crvwaveforminfo branches"), false};
+        fhicl::Atom<bool> fillcrvcoincs{Name("FillCRVCoincs"),Comment("Flag for turning on crv CoincidenceClusterbranches"), false};
+        fhicl::Atom<bool> fillcrvpulses{Name("FillCRVPulses"),Comment("Flag for turning on crvpulses(mc), crvdigis branches"), false};
```
Similar to comment to earlier about whether we want to rename this parameter FillCRVDigis

sam-grant replied:
See my reply to the earlier comment: maybe we should split `digis` and `pulses` up and have two separate flags.

AndrewEdmonds11 commented on src/TrkAnaTreeMaker_module.cc:
```diff
@@ -264,19 +264,19 @@ namespace mu2e {
-      std::vector<CrvWaveformInfo> _crvwaveforminfo;
-      std::vector<CrvHitInfoMC> _crvpulseinfomc;
+      std::vector<CrvPlaneInfoMC> _crvcoincsmcplane;
+      std::vector<CrvPulseInfoReco> _crvpulses;
+      std::vector<CrvWaveformInfo> _crvdigis;
```
I'm afraid I'm now confused as to the difference between pulses, waveforms, and digis...

sam-grant replied:
I agree it's confusing, I had to ask Yuri. 

`digis` and `waveforms` are the same thing, they're ADU counts binned in 12.5 ns intervals. `pulses` are reconstructed distributions of `digis` (fitted and integrated).

AndrewEdmonds11 replied:
Ah, OK let's move to just using ```digis`` then and try to remove ```waveform``` everywhere we can

AndrewEdmonds11 reviewed (changes requested):
Thanks, Sam - this is great! I've left a few comments just to improve things. One thing I did get confused about is the difference between digis, pulses, and waveforms. Could you clarify that for me?

brownd1978 commented:
std::array<> is pin-compatible with c-arrays, plus provide copy constructor
(which I guess is what Rob was getting at).

On Mon, Feb 26, 2024 at 9:49 AM Rob Kutschke ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In inc/CrvHitInfoReco.hh
> <https://github.com/Mu2e/TrkAna/pull/124#discussion_r1503064756>:
>
> > @@ -32,7 +33,10 @@ namespace mu2e
>        nHits(nCoincidenceHits),
>        nLayers(nCoincidenceLayers),
>        angle(coincidenceAngle)
> -    {}
> +    {
> +      // Copy PEsPerLayer array elements
> +      for (int i = 0; i < 4; i++) this->PEsPerLayer[i] = PEsPerLayer[i];
>
> Is it too late to define all of the c-style arrays of dimension 4 as
> std::array<int,4>? Then you can do the copy in the intiailizer list:
> PEsPerLayer(PEsPerLayer) . One of the obscure but handy corners of c++ is
> that in an initializer list, the compiler knows that it means to have a
> data member and an argument with the same name
>
> It it is too late definitely use std::array, then use CRVId::nLayers from
> Offline/DataProducts/inc/CRVId.hh . That's a compile time condstant rather
> than a variable that you have to look up at run time.
>
> I don't think that the recommendation of
> Offline/GeometryService/inc/CosmicRayShieldMaker.hh will work. Do not try
> to make an instance of that class yourself. It is to only be using insde
> the geometry service.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/TrkAna/pull/124#discussion_r1503064756>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH576CV6KBRUPOBOVD33TYVTDJPAVCNFSM6AAAAABDXFGRK2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSMBRGYYTQNJTGA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
David Brown            ***@***.***
Office Phone (510) 486-7261
Lawrence Berkeley National Lab
M/S 50R5008 (50-6026C)  Berkeley, CA 94720

kutschke commented:
> std::array<> is pin-compatible with c-arrays, plus provide copy constructor (which I guess is what Rob was getting at).

Yes.  That's what I was getting at.

Sam: is there a good reason not to convert to std::array?

sam-grant commented:
No, not that I’m aware of. I’m testing the std::array implementation now.

From: Rob Kutschke ***@***.***>
Date: Monday, February 26, 2024 at 14:38
To: Mu2e/TrkAna ***@***.***>
Cc: Grant, Samuel ***@***.***>, Author ***@***.***>
Subject: Re: [Mu2e/TrkAna] Updated CRV branch names and added a PEs per layer variable. (PR #124)

std::array<> is pin-compatible with c-arrays, plus provide copy constructor (which I guess is what Rob was getting at).

Yes. That's what I was getting at.

Sam: is there a good reason not to convert to std::array?

—
Reply to this email directly, view it on GitHub<https://github.com/Mu2e/TrkAna/pull/124#issuecomment-1965225977>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKP54AWZGSJBUGRDGUFJHUTYVTXGDAVCNFSM6AAAAABDXFGRK2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRVGIZDKOJXG4>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

sam-grant commented:
> No, not that I’m aware of. I’m testing the std::array implementation now. std::array<> is pin-compatible with c-arrays, plus provide copy constructor (which I guess is what Rob was getting at). Yes. That's what I was getting at. Sam: is there a good reason not to convert to std::array? — Reply to this email directly, view it on GitHub<[#124 (comment)](https://github.com/Mu2e/TrkAna/pull/124#issuecomment-1965225977)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKP54AWZGSJBUGRDGUFJHUTYVTXGDAVCNFSM6AAAAABDXFGRK2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRVGIZDKOJXG4>. You are receiving this because you authored the thread.Message ID: ***@***.***>

I've pushed my changes where I use std::array.

sam-grant commented:
I think I've addressed all the comments now. The only one that I'm not sure about is commit [140f9a9](https://github.com/Mu2e/TrkAna/pull/124/commits/140f9a9211a2e26abaa915310d883174f5c0adf4): it works but I'm not sure if it's an improvement on what was already there, maybe it is less confusing now? Please let me know. Thanks.

kutschke commented on inc/CrvInfoHelper.hh:
```diff
@@ -60,7 +66,25 @@ namespace mu2e
+        for(iter=counters.begin(); iter!=counters.end(); iter++)
+        {
+          const CRSScintillatorBarIndex &barIndex = (*iter)->index();
+          for(int SiPM=0; SiPM<4; SiPM++)
```
You should also use a proper identifier for this literal 4 and for the literal 4 a few lines down.  This happens to be a different 4 than one you addressed earlier today.  I believe that this 4 is CRVId::nChanPerBar, (which is only accidentally the same value as CRVId::nLayers).  This is one of the reasons we want people to use named constexpr or variables, not numerical literals.

@ehrlich-uva please let us know whether or not my 4-ology is correct.

@ehrlich the construction ( something.barIndex.asInt()*4 +SIPM) appears several times in this code.  And I bet it appears in other places.  Do you provide a helper function that wraps this calculation?  If so, where is it?

@sam-grant if Ralf has such a function, please use it.  If not leave it for another day.

sam-grant replied:
Hi Rob. I agree that this is a different 4 to the number of layers, and I agree that `CRVId::nChanPerBar` looks like the right number to use, since there are 4 SiPMs per scintillator bar. @ehrlich-uva can you confirm? Thanks.

sam-grant replied:
I see that this calculation also appears in `CrvInfoHelper.cc` as you said, I'll change that to use the identifier as well.

sam-grant replied:
I just pushed the change from `4` to `CRVId::nChanPerBar`.

kutschke replied:
Thanks.  I will approve it now.

AndrewEdmonds11 reviewed (approved):
All looks OK to me. Thanks, Sam :)

kutschke commented:
> I think I've addressed all the comments now. The only one that I'm not sure about is commit [140f9a9](https://github.com/Mu2e/TrkAna/pull/124/commits/140f9a9211a2e26abaa915310d883174f5c0adf4): it works but I'm not sure if it's an improvement on what was already there, maybe it is less confusing now? Please let me know. Thanks.

Thanks for making the changes.  I don't know enough about what the code is doing to comment on this question.

I noticed another case of an inappropriate use of a numeric literal and requested a change.

oksuzian commented:
@AndrewEdmonds11, when do you plan to merge this PR?

AndrewEdmonds11 commented:
Hi Yuri - thanks for checking in. I'm planning to merge it in before the Collaboration Meeting tutorial so that it is part of the first v05 release.
